### PR TITLE
[ZEPPELIN-1876] improved comptetion with schema/table/column separati…

### PR DESCRIPTION
**Pull request by sotnich**
…on + sqlcompleter tests

### What is this PR for?
This PR changes autocompletion behaviour in jdbc interpeter.
There are some changes:
* [main change] autocompletion now depends on what are you typing. Now there are four types of competion: schema, table, column and keywords. If you typing new word then autocompetion suggests only keywords and schema names. If you are typing after schema name with point then you get list of tables in that schema. Also if you typing a name after point after a table name you will get a list of column names of this table.
* autocomption now supports aliases in sql. If you write alias for a table in sql you will get a list of columns for an aliased table if you write down alias and point.
* autocompletion now load keywords only in low case (otherwise there are so many keywords in a list of autocompletion that it is becomes uncomfortable)

### What type of PR is it?
Improvement

### Todos
* [ ] - sort names in the output of autocompletion
* [ ] - list only schema names if we are typing a schema name (for example after keywork FROM)
* [ ] - add description in autocompletion list for schema names - schema, for table names - table and so
* [ ] - autocompletion must initialize on opening of interpeter (not only after execution of sql)
* [ ] - update autocompletion schemas only after execute update sql, not after every sql ???
* [ ] - new option for postgresql interpreter: postgresql.completer.schema.filter. Filter schema names loaded into autocompletion (no more garbage schema names).

### What is the Jira issue?
* Open an issue on Jira https://issues.apache.org/jira/browse/ZEPPELIN/
* Put link here, and add [ZEPPELIN-*Jira number*] in PR title, eg. [ZEPPELIN-533]

### How should this be tested?
Outline the steps to test the PR here.

### Screenshots (if appropriate)
https://issues.apache.org/jira/secure/attachment/12845228/auto1.JPG
https://issues.apache.org/jira/secure/attachment/12845229/auto2.JPG
https://issues.apache.org/jira/secure/attachment/12845230/auto3.JPG

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes

### CODE REVIEW

**Comment by jongyoul** [comment]
@astroshim Could you please review this PR?

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java [hunk 1]
```diff
@@ -51,84 +51,143 @@
...
+          String schemaName = schemas.getString("TABLE_SCHEM");
+          if (schemaFilter.equals("") || schemaFilter == null || schemaName.matches(
+                  schemaFilter.replace("_", ".").replace("%", ".*?"))) {
+            res.add(schemaName);
+          }
+        }
+      } finally {
+        schemas.close();
+      }
+    } catch (Throwable t) {
```
Comment (felixcheung) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java]: Is Exception ok? Should it be Throwable?

Comment (sotnich) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/SqlCompleter.java]: Yeah, thanks, SQLException is better. Have changed it.

**Comment by sotnich** [comment]
Don't think CI errors are related to this PR.

File: jdbc/src/test/java/org/apache/zeppelin/jdbc/SqlCompleterTest.java [hunk 2]
```diff
@@ -0,0 +1,313 @@
+package org.apache.zeppelin.jdbc;
```
Comment (jongyoul) [inline jdbc/src/test/java/org/apache/zeppelin/jdbc/SqlCompleterTest.java]: You need to add license header into this file

Comment (sotnich) [inline jdbc/src/test/java/org/apache/zeppelin/jdbc/SqlCompleterTest.java]: Fixed, but still there are erros that I don't understand how to link to PR.

Comment (jongyoul) [inline jdbc/src/test/java/org/apache/zeppelin/jdbc/SqlCompleterTest.java]: @sotnich All .java files should have ASF license header on top of each file. You didn't add it and you can add it by copying the header from one of other java files. :-)

**Comment by jongyoul** [comment]
There is a flaky test in the result:

```
WelcomePageSuite
```

Never mind the result. that test is being fixed now. LGTM

**Comment by jongyoul** [comment]
@sotnich Can you rebase it onto current master? Recently, we fixed some flaky tests including your case. It would make your CI green

**Comment by sotnich** [comment]
Made a rebase, now CI tests have only one error related with DistributedResourcePoolTest.
Think it's still not related to my PR.

**Comment by cloverhearts** [comment]
@sotnich 
I was test.
LGTM.

I have a question.
Is it possible to only support pgsql? (actually, i am mysql/maria user)

**Comment by sotnich** [comment]
@cloverhearts 
Please, clirify your question.
This PR supports every JDBC driver that can return schema and table names.
I have tested it on postgres, but I am sure it would correctly run on mysql and other relational databases.

**Comment by astroshim** [comment]
@sotnich Thank you for great improvement.
I tested `mysql` and `postgresql` and only `postgresql` is working expectedly like @cloverhearts.

**Comment by Leemoonsoo** [comment]
CI failure is not related to this change.
LGTM and merge to master if no further discussion.

**Comment by sotnich** [comment]
@cloverhearts @astroshim 
Made a little change - now mysql also works as well as postgresql. I have tested it.

**Comment by astroshim** [comment]
@sotnich Great Job! but still doesn't work with `mysql`.

**Comment by cloverhearts** [comment]
auto-com

Hello, I was check again.
but, does not work on mysql.
Is there anything I missed?

**Comment by sotnich** [comment]
@cloverhearts @astroshim, check it now, please. If it's still not working send me details: configuration of the interpreter and jdbc log file.

**Comment by cloverhearts** [comment]
@sotnich 
now Well working!
good works!
Thank you for great feature!

mysql

**Comment by Leemoonsoo** [comment]
LGTM and merge to master if no more discussions.

**Pull request closed without merging**